Products/CMFPlone/CatalogTool.py

```diff
@@ -379,6 +379,51 @@ def getCounter(self):
         processQueue()
         return self._counter is not None and self._counter() or 0
 
+    @security.private
+    def allow_inactive(self, query_kw):
+        """Check, if the user is allowed to see inactive content.
+        First, check if the user is allowed to see inactive content site-wide.
+        Second, if there is a 'path' key in the query, check if the user is
+        allowed to see inactive content for these paths.
+        Conservative check: as soon as one path is disallowed, return False.
+        If a path cannot be traversed, ignore it.
+        """
+        allow_inactive = _checkPermission(AccessInactivePortalContent, self)
+        if allow_inactive:
+            return True
+
+        paths = query_kw.get('path', False)
+        if not paths:
+            return []
+
+        if isinstance(paths, dict):
+            # Like: {'path': {'depth': 0, 'query': ['/Plone/events/']}}
+            # Or: {'path': {'depth': 0, 'query': '/Plone/events/'}}
+            paths = paths.get('query', [])
+
+        if isinstance(paths, basestring):
+            paths = [paths]
+
+        objs = []
+        site = getSite()
+        for path in list(paths):
+            path = path.encode('utf-8')  # paths must not be unicode
+            try:
+                objs.append(site.restrictedTraverse(path))
+            except (KeyError, AttributeError):
+                # When no object is found don't raise an error
+                pass
+
+        if not objs:
+            return False
+
+        allow = True
+        for ob in objs:
+            allow = allow and\
+                _checkPermission(AccessInactivePortalContent, ob)
```

Comment:
Yes, that's indented - I think it's better to be conservative here instead of showing too much content which shouldn't be shown.
Also, it's quite uncommon that two paths are used.